packages/logging/src/logging-and-telemetry.ts

```diff
@@ -0,0 +1,785 @@
+/* eslint prefer-arrow-callback: "error" */
+import redactInfo from 'mongodb-redact';
+import { redactURICredentials } from '@mongosh/history';
+import type {
+  MongoshBus,
+  ApiEventWithArguments,
+  ApiEvent,
+  UseEvent,
+  EvaluateInputEvent,
+  ShowEvent,
+  ConnectEvent,
+  ScriptLoadFileEvent,
+  StartLoadingCliScriptsEvent,
+  StartMongoshReplEvent,
+  GlobalConfigFileLoadEvent,
+  CryptLibrarySkipEvent,
+  CryptLibraryFoundEvent,
+  SnippetsCommandEvent,
+  SnippetsErrorEvent,
+  SnippetsFetchIndexErrorEvent,
+  SnippetsFetchIndexEvent,
+  SnippetsLoadedEvent,
+  SnippetsLoadSnippetEvent,
+  SnippetsNpmDownloadActiveEvent,
+  SnippetsNpmDownloadFailedEvent,
+  SnippetsNpmLookupEvent,
+  SnippetsRunNpmEvent,
+  SnippetsTransformErrorEvent,
+  EditorRunEditCommandEvent,
+  EditorReadVscodeExtensionsDoneEvent,
+  EditorReadVscodeExtensionsFailedEvent,
+  FetchingUpdateMetadataEvent,
+  FetchingUpdateMetadataCompleteEvent,
+  SessionStartedEvent,
+  MongoshBusEventsMap,
+} from '@mongosh/types';
+import { inspect } from 'util';
+import { MongoLogWriter } from 'mongodb-log-writer';
+import { mongoLogId } from 'mongodb-log-writer';
+import type {
+  AnalyticsIdentifyMessage,
+  MongoshAnalytics,
+  MongoshAnalyticsIdentity,
+} from './analytics-helpers';
+import { hookLogger } from '@mongodb-js/devtools-connect';
+import { MultiSet, toSnakeCase } from './helpers';
+import { Writable } from 'stream';
+
+export class MongoshLoggingAndTelemetry {
+  log: MongoLogWriter;
+  pendingLogEvents: CallableFunction[] = [];
+
+  telemetryAnonymousId: string | undefined;
+  userId: string | undefined;
+  trackingProperties: {
+    mongosh_version: string;
+    session_id: string;
+  };
+  userTraits: AnalyticsIdentifyMessage['traits'] & {
+    [key: string]: unknown;
+  };
```

Comment:
Let's mark the things that are readonly here with `readonly`/`Readonly<>` to make it a bit easier to follow what's state of the class and what's just parameters.